### ISSUE TITLE
save/reload data and weights

### ISSUE THREAD

**Issue by danjdrennan** [katzfuss-group/batram-cov#1]
The getting started notebooks demonstrate how to create a `train_module` and use
it to instantiate a `model = CovariateTransportMap(train_module, ...)`. The
`train_module` contains both the training data and the parameters for a sequence
of regression problems. In one application we are interested to keep the `model`
handle fixed but update either

- the parameters in `train_module`
- or update the data in `train_module`.

Both are necessary, and we would like to be able to fully restore a trained
model from disk. Currently that is impossible due to the abstractions we have
written, particularly because of a local function that we cannot serialize
through when saving graphs and their corresponding data.

**Comment by danjdrennan** [comment]
Leaving the issue open for further discussion. A few prototype methods were merged to `src/batram_cov/model.py::CovariateTransportMap` in commits:

- a4ebdc6 model(save/load weights): save and load data using ocp.checkpoints
- 2550813 model(restore states): helper to restore parameters from a previous training state
- 97db8fc model: docs for update training data function
- d8a43f5 proto: add new x and y to model
- f94c89e proto: data updating functions

It is **highly recommended** to read the source before using these methods, as the interfaces are not well tested. The functions are

```python
class CovariateTransportMap:
    # commit: a4ebdc6 model(save/load weights): save and load data using ocp.checkpoints
    def save_model(self, abs_path, checkpointer=None):
        ...

    # commit: a4ebdc6 model(save/load weights): save and load data using ocp.checkpoints
    def load_model(self, abs_path, checkpointer=None):
        ...

    # commits:
    # - 97db8fc model: docs for update training data function
    # - d8a43f5 proto: add new x and y to model
    # - f94c89e proto: data updating functions
    def update_train_data(self, new_x, new_samples):
        ...

    # commit: 2550813 model(restore states): helper to restore parameters from a previous training state
    def update_model_params(self, previous_state, use_data=False):
        ...
```

We are using `flax.nnx` to store models and weights. Any `nnx.Module` can be split into `graph, state = nnx.split(module)`, which basically creates a PyTree of object definitions (in `graph`) and object state (in `state`). The states are dictionary-like objects with similar semantics for looping over objects and inserting the correct data.

We don't do any form of tree walking in our updates, assuming you either want to update the entire parameter node or that you want to update all of the items in a data module. In all cases, this is just walking across the key-value pairs of one object and copying their data into another object.

**Comment by anirban0451** [comment]
<img width="1258" height="576" alt="Image" src="https://github.com/user-attachments/assets/7970402c-4dbd-456d-bcb9-e64918b06bde" />

I am trying to update the training dataset by appending new datasets in the getting-started notebook file. For example, in this particular example, I am trying to add 1 sample of shape (1, 1, 256). The usual training dataset is of shape (1, 5, 256). On debug, I find that, `nn_idx` is not able to properly subset. Same issue persists when I reshape the data to size (1, 256). However, when I just pass a one-dimensional object as a new sample, even though the error persists, it's definition changes.

<img width="1010" height="579" alt="Image" src="https://github.com/user-attachments/assets/b243900d-3985-4fbf-b249-be544d81e298" />

Upon debugging, I found that `.transpose()` is throwing the error, as the resulting subset is only 2-dimensional, and not 3-dimensional. Do you think a bug might be present and/ or I am missing something?

**Comment by danjdrennan** [comment]
@anirban0451, thanks for sharing the note. The original interface required that
you pass `new_samples` with shape `(num_locs, num_samples)`. That is
understandably confusing considering the rest of the interfaces, so I've
transposed that function.

The updated function has the following docs:

```python
# From batram_cov.model
class CovariateTransportMap:

    def update_train_data(self, new_x: np.ndarray, new_samples: np.ndarray):
        """Add new_x and new_samples to training data module

        NOTE: The function assumes all necessary preprocessing has been conducted by
        the user _before_ calling this function. This function only concatenates data
        into the training module state.

        Args:
        new_x: an x value with shape (num_samples, x_dim)
        new_samples: fields with shape (num_samples, num_locs)

        Returns:
        None
        """
        ...
```

For a quick test, I adapted the `getting-started.ipynb` code to test if this
works whenever I want to add new data to the training data module like so:

```python
# Example building `valid_data` in getting-started.py:
valid_data = jax.tree.map(
    lambda a: jax.device_put(a),
    [ locs[maxmin_ordering], nns, li, maxmin_ordering, valid_x, valid_y[..., maxmin_ordering], ],
)
valid_data = Data(*valid_data)

# Intermediate model fitting steps skipped ...

# NOTE: The valid_data should have produced with this shape
assert valid_data.samples.shape == (1, 49, 256), f"{valid_data.samples.shape=}"

# Now update the model with the validation set
model.update_train_data(valid_data.x, valid_data.samples[0])
```

The `assert` statement in that snippet shows the shape of the data, which
matches the dimensionality you are working in. The indexes of that array are
`(num_reps, num_samples, num_locs)`. If you're using the same interfaces as we
generated for the `getting-started.ipynb` then you should be indexing in exactly
the same way as above.

Please let me know if this has any other problems. We can keep the issue open
for now as we keep working on that interface.

**Comment by anirban0451** [comment]
I tried optimizing the model once new data is loaded. The optimization is running smoothly, which is great! 💯 

However, I think the current setup would not work well if I have more than one sample per new x, right? I think the current `model.update_train_data()` assumes that we have one sample per new x. Please correct me if I am wrong.

**Comment by danjdrennan** [comment]
@anirban0451, that's correct. It's unfortunately impossible to introspect the
inputs and do something sensible for users in all cases. The easiest thing to do
is to call `model.update_train_data` multiple times, like

```python
# Something like this, but it will depend on how your data are shaped of course
model.update_train_data(new_x, new_y[0])
model.update_train_data(new_x, new_y[1])
```

A more elaborate approach that could work is to do something like
stacking/tiling the data before calling `model.update_train_data`. That could
look like

```python
new_x = jnp.stack([x for _ in range(y.shape[0] // x.shape[0])], axis=0)
new_x = new_x.reshape(-1, x.shape[-1])

model.update_train_data(new_x, new_y)
```

If the `new_y` data are shaped like `new_y.shape == (num_reps, num_samples, N)`
then you'd want to fold the first two dimensions together using
`new_y.reshape(-1, y.shape[-1])` before passing them into the update function.

Note these all assume that the data should be stacked rather than tiled. I think
our generating processes would all satisfy that assumption, but you could end up
with bad training runs if this is the wrong alignment. Double check this before
implementing any abstractions in your application.

An additional concern that I need to follow up on is investigating how we use
the `normalize_x` flag in this code:

```python
def create_and_merge_problems(
    data: ArraysProtocol,
    num_ip_f: int,
    num_ip_g: int,
    sample_idx: int = 0,
    max_nns: int = 30,
    whiten: bool = False,
    variational_noise_f: float = 0.0,
    variational_noise_g: float = 0.0,
    ip_fixed: bool = True,
    ip_distance_penalty: None | float = None,
    linear_only: bool = True,
    normalize_x: bool = True,
) -> rp.HGPIPProblem:
    ...
```

I'm pretty sure we calculate the normalizing parameters once and then apply them
as needed in the model. However, for what you're doing, I think it would be
smarter to disable that flag and handle normalization yourself outside of the
`model` object. I'll look into this further but it's something we should
potentially discuss together when you have an opportunity to.

Dan

**Comment by danjdrennan** [comment]
# Normalizing x values

Following up on the note about normalizing x values. A sketch of what we do
specifically is the following (this is idealized to remove some complexity):

```python
class ModelComponent:
    def __init__(self, data, normalize_x):
        self.x_mean = data.covariates.mean()
        self.x_std = data.covariates.std()
        self.normalize_x = normalize_x

    def _get_x(self, x):
        if self.normalize_x:
            return (x - self.x_mean) / self.x_std

        return x
```

The rest of this comment is separated into a few subsections.

## Facts about current implementation

To start, the following two facts are true about our implementation of the model
and this patched `CovariateTransportMap.update_train_data` method:

1. We do not modify any data when storing it. It only gets normalized in the
   computation graph, not when creating the objects.

2. Updating the training data through `CovariateTransportMap.update_train_data`
   does not reparameterize `x_mean` or `x_std`.

3. The `x_mean` and `x_std` are viewed as additional facts about the data, not
   as parameters of the model. This means it is relatively easy to update them
   in `update_train_data` if we want to.

4. When you update the train data, it will get normalized in the model with the
   same `x_mean` and `x_std` as all previous data were normalized by. (Again,
   this happens every time we call the model.)

## Changing scale

As shown, we calculate the mean and standard deviation using simple arithmetic
statistics. That works well when the number of x values are the same, but might
not be a good choice when there are several repeated observations.

Basically my concern is that the model you've already fit is tuned to whatever
scale the `x` values were on and the simple normalization we may have computed
in the first place. If we update `x_mean` and `x_std` using the new data then it
becomes unclear how the scale of the `x` values changes over time, especially if
we're not weighting the observations somehow.

Three simple cases to consider. In all three cases, suppose wlog that the
original data are $x_1, \dots, x_n \sim N(0, 1)$.

**Example 1**: Assume the new data are $x_1', \dots, x_m' \sim N(0, 1)$.
In this case we are adding new observations from the same scale as the
initial data, and updating $\bar x$ and $s$ probably has no noticeable effect on
the normalization. Moreover, the new combined data are on the same scale as the
model was previously trained on.

**Example 2**: Suppose we insert $x_1', \dots, x_n' \sim N(5, 1)$, 
where $n$ is the same size as before. If we recompute the mean and standard
deviation with this new data, then in expectation it has $\bar x = 2.5$ and $s =
2.7$. Basically now we've created a mixture distribution in covariates, and the
scale of this renormalized data are going to have no meaningful relationship to
the original scale of the data we started with.

**Example 3**: Now assume we insert new data $x_1', \dots, x_m' \sim N(0, \epsilon)$
for $m < n$ and $0 < \epsilon < 1$. Now in expectation the
mean is $\bar x = 0$ but the standard deviation is $s = \sqrt{m \text{var}
\epsilon^2 + n} < 1$. Depending on the size of $\epsilon$ and the number of new
observations $m$, this could concentrate anywhere in the unit interval. Again,
this would make the original scale that the model learned kind of meaningless,
and retraining to learn the new scale may be extensive.

Other examples that are more complex, but closely related to **Example 3** would
involve putting repeated observations into the problem.

## How to update the normalization

On one hand, if you ask to normalize the data when constructing the model then I
think it is reasonable for our interface to update those normalizing values for
you. This is additional work, but I think we could do it more easily inside the
interface than you can managing it externally as a user. The drawback if you're
making several updates to the model data is that the scale of that normalization
can become severely distorted in just a few iterations.

Taking all of these things together with my understanding of your intended use
case, I would recommend you manage data normalization yourself externally so
that you can manage/monitor the scale of the data more clearly than it would be
in our black box.

As to an interface, we could add flags like `renormalize_x` and
`weight_normalization` to all of the points where we implement normalization.
That would give more explicit control over the case where you're inserting
imbalanced data either at the start or later when you want to insert more data
into the training pipeline. Those flags seem like good ones to add to the
implementation regardless, but I think for your application we should discuss
them in more detail.

**Comment by danjdrennan** [comment]
Added prototypes of these features in:

- a4ebdc6 model(save/load weights): save and load data using ocp.checkpoints

- 2550813 model(restore states): helper to restore parameters from a previous training state

- 97db8fc model: docs for update training data function

They are not well tested and may need to be improved from the current state.
